### ISSUE TITLE
Tracking Issue for `wasi_ext`

### ISSUE THREAD

**Issue by coolreader18** [rust-lang/rust#71213]
The feature gate for the issue is `#![feature(wasi_ext)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also uses as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

When should this be stabilized? Is wasi as a specification stable enough for wasi-specific APIs to be included in std? 

### Implementation history

#59619

**Comment by bstrie** [comment]
Does anyone know what the status of this feature is?

**Comment by a1phyr** [comment]
I don't think we want to stabilize this feature while WASI is not stable itself.

**Comment by alyssais** [comment]
The methods in `std::os::wasi::fs::FileExt` look a lot like the POSIX `*at` methods.  Would it be possible to make this extension available for all Unix platforms, rather than limiting it to wasi?

There are a lot of reasons to want to use these methods from other Unix operating systems:

- They make it possible to work with paths longer than MAX_PATH.
- They are the only way to interact with files in a Capsicum sandbox on FreeBSD.
- They can prevent TOCTOU vulnerabilities where a directory is moved between accesses to files within that directory.

**Comment by banocean** [comment]
> I don't think we want to stabilize this feature while WASI is not stable itself.

Are we talking about WASI 1.0 or including following APIs in WASI specification ([which just happened as part of WASI Preview 2](https://github.com/WebAssembly/WASI/blob/main/preview2/README.md#wasi-preview-2-contents))?

**Comment by thaliaarchi** [comment]
Following https://github.com/rust-lang/rust/pull/137165 (Use `tell` for `<File as Seek>::stream_position`), `tell` is now directly exposed via `stream_position`, making `<File as FileExt>::tell` redundant. I propose to remove it in https://github.com/rust-lang/rust/pull/137205.

**Comment by eduardomourar** [comment]
Concrete downstream consumer, in case it's useful signal for prioritization: [uutils/coreutils#13625](https://github.com/uutils/coreutils/pull/13625) needed `MetadataExt::ino`/`dev`/`nlink` and `std::os::wasi::fs::symlink` on `wasm32-wasi` and, lacking stable access, worked around it with a direct `rustix::fs::stat`/`symlink` dependency instead (see `uucore::fs::FileInformation` and `cp::symlink_file` in that PR). A stable subset of `wasi_ext` (at least the `MetadataExt` inode/device accessors) would let it drop that workaround and match the Unix code path structurally.

**Comment by a1phyr** [comment]
I'd like to add that we should probably stabilize most of the API, as wasip1 won't change again at this point.
